doc/README.twom.md

```diff
@@ -0,0 +1,189 @@
+
+Along with a three phase commit, this makes the twoskip file format incredibly robust.  The three phases are:
+
+1. before writing any new record to the end of file, mark the header as DIRTY, and fsync
+2. once all new records are written, fsync the entire file (still marked dirtry)
+3. mark the header NOT-DIRTY along with the new file length, and fsync
+
+This means that so long as the final header write is atomic, transactions are always completely safe.
+
+## Twoskip downsides:
+
+- Three fsyncs per transaction is kinda expensive!
+- Every write causes the file to be extended, as twoskip format uses exact file size and doesn’t allocate slop.
+- A single write does amplify into 1+N (the level of the record) writes; the append of the new record, and the updates of the immediately previous records in the skiplist at each level.  Since half the records are level 1, half that level 2, etc - the average write updates three separate previous records.  In a recently repacked file, the previous records are likely to be adjacent or nearly so for the lower levels, but this becomes less so over time, so repacks are valuable for both data locality and size reduction.
+- The repack operation requires the database to be remain exclusively locked for the duration, which is slow on large files (particularly along with the above lack of slop).
+
+## The Twom format
+
+Why the name?  It’s twoskip, but with mmap and MVCC.  Two Ms.  And the “tombstone” records, rather than just causing the skiplist to be rewritten without the deleted record in it, remain in the linked lists right until a repack, with an ANCESTOR pointer back to the previous entity with that key.
+
+This tombstone allows the MVCC transaction, because you can walk the current lowest level linked list, and any time you find a record with an offset in the future of the file size at the time you started the transaction, walk back up the ancestor chain (which is a single linked list backward through the file, to the original ADD record which was the first time the key was seen in this file).
+
+With MVCC transactions, we can then re-implement the repack operation; instead of having to lock the file for the entire repack, we can take a read-only lock and yield it many times while walking the original records and adding them to the new file in sorted order, and then (still releasing locks as many times as we like) read the remainder of the file and replay the operations (adds, replaces and deletes) to the new file.  This doesn’t require any exclusive lock at any point, so readers can continue throughout the repack, even until the rename - and either get the old file or the new file.
```

Comment:
If I read this correctly, then the repack won't result in an optimally packed file, e.g. it might include tombstones and ancestors. The determining factors seem to be the number of read-only lock yields by the repacker and the number of writes between those yields. Is there some heuristic to choose how often to yield the read-only locks when the average number of writes is known?

Reply:
Strictly, repack will be optimally repacked for everything in the file that was present at the time the repack started, so for sure that much will be optimal (and it was already so poor at that point that it calculates it will save a bunch of space) - so it'll definitely be better than not repacking, but could still be non-ideal if there's a lot of writes while the repack is taking place.

The heuristic is to release 64 times less often when repacking than in a usual 'foreach' to make repack a bit more greedy.  So that's every 64k records (usually it releases every 1024 records)